Add unregister() via DefaultQISKitProvider changes

diego-plan9 opened the pull request:
### Summary

Add `wrapper.unregister()` for removing previously registered providers,
by delegating into `DefaultQISKitProvider` internal tracking of
provider instances:
* use a dict instead of a list for `self.providers`, introducing the
  ability to assign a user name to a provider instance at the wrapper
  level.
* update `register(...,provider_name)` parameter so it is set by the
  user explicitly (with "weak" fallbacks for the two main use cases,
  for backwards compatibility).
* add `wrapper.registered_providers()` for convenience, displaying the
  list of currently registered providers.

Regroup wrapper tests, adding basic ones for the new functionality and
fixing an issue with global variables and imports via `tearDown`.

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Details and comments

Even if a related functionality was included in #540, after discussion with Jay I'm moving this feature to its own PR for being able to incorporate it into the codebase independently and hopefully cater to the urgency, as #540 is ongoing a refactoring and has a wider scope. The modifications for `DefaultQiskitProvider` are actually part of that refactor - once this PR is merged, changes will be moved onto the other pull request.

delapuente commented on test/python/test_wrapper.py:
```diff
@@ -0,0 +1,93 @@
+                                provider_name='provider1')
+        qiskit.wrapper.register(QE_TOKEN, QE_URL, hub, group, project,
+                                provider_name='provider2')
+        self.assertEqual(['local', 'provider1', 'provider2'],
```
Non-blocking: I'd relax the test and simply test if the names are in `registered_providers()`.

diego-plan9 replied:
Good catch - the tests needed a revamp! I'll make them to operate on the "relative" difference of providers registered at every point in time, in order to avoid having to take into account `local`.

delapuente commented on test/python/test_wrapper.py:
```diff
@@ -0,0 +1,93 @@
+
+    def test_register_unknown_name(self):
+        """Test registering a provider with not explicit name."""
+        with self.assertRaises(QISKitError):
```
Non-blocking: I'd relax this constraint by asserting the list before and after the intent to register are equal:

```python
before_attempt = qiskit.wrapper.registered_providers()
with self.assertRaises(QISKitError):
  qiskit.wtapper.register('FAKE_TOKEN', 'http://unknown')
self.assertEqual(sorted(qiskit.wrapper.registered_providers()), sorted(before_attempt))
```

This way you're checking that a failing operation does not alter the list of providers, instead of tying yourself to a particular implementation such as the `'local'` name.

delapuente commented on test/python/test_wrapper.py:
```diff
@@ -0,0 +1,93 @@
+    @requires_qe_access
+    def test_unregister(self, QE_TOKEN, QE_URL, hub, group, project):
+        """Test unregistering."""
+        qiskit.wrapper.register(QE_TOKEN, QE_URL, hub, group, project,
```
Same here, we could relax the test by replacing the last assertion with a comparison between the current `registered_providers` and those before registering anything and...

delapuente commented on test/python/test_wrapper.py:
```diff
@@ -0,0 +1,93 @@
+        """Test unregistering."""
+        qiskit.wrapper.register(QE_TOKEN, QE_URL, hub, group, project,
+                                provider_name='provider1')
+        self.assertEqual(['local', 'provider1'],
```
...replacing this one with an inclusion assertion.

delapuente commented on test/python/test_wrapper.py:
```diff
@@ -0,0 +1,93 @@
+
+    def test_unregister_non_existent(self):
+        """Test unregistering a non existent provider."""
+        with self.assertRaises(QISKitError):
```
Same here.

delapuente commented on CHANGELOG.rst:
```diff
@@ -20,6 +20,7 @@ The format is based on `Keep a Changelog`_.
 
 Added
 -----
+- Add ``unregister()`` for removing previously registered providers (#584).
```
If changing the signatures or `register()` and `unregister()` for them to return the provider instances, consider adding that here too.

diego-plan9 commented on qiskit/wrapper/defaultqiskitprovider.py:
```diff
@@ -76,19 +78,79 @@ def deprecated_backend_names(self):
         """
-        self.providers.append(provider)
+        if provider_name == 'local':
+            raise QISKitError("Cannot unregister 'local' provider.")
```
Yes, it's very much up for debate - it was mostly based on https://github.com/QISKit/qiskit-core/pull/547/files/1ca64661029eba38d25edcc5c1f0ba61388c9581#diff-c8057d22bae58d0abb388944221cc41d. I'd be happy to re-discuss it when the other PR is addressed, as I don't have a strong opinion either way!

diego-plan9 commented on qiskit/wrapper/defaultqiskitprovider.py:
```diff
@@ -76,19 +78,79 @@ def deprecated_backend_names(self):
+        Args:
+            credentials_dict (dict): dictionary of credentials for a provider.
+            provider_name (str): User name for the provider. A name will
+                automatically
```
Yep :man_facepalming:

diego-plan9 commented on qiskit/wrapper/defaultqiskitprovider.py:
```diff
@@ -76,19 +78,79 @@ def deprecated_backend_names(self):
+                'A provider with name "%s" is already registered.'
+                % provider_name)
+
+        self.providers[provider_name] = provider
```
Sure, makes sense!

diego-plan9 commented on qiskit/wrapper/_wrapper.py:
```diff
@@ -38,17 +37,33 @@ def register(token, url='https://quantumexperience.ng.bluemix.net/api',
+    Raises:
+        QISKitError: if the provider name is not valid.
+    """
+    _DEFAULT_PROVIDER.remove_provider(provider_name)
```
Hmm, I'm unsure if it would play nicely with the philosophy of not having the end-user-oriented wrapper functions expose the providers directly (and in this particular case, would return a "not longer in use provider"). It's a good point, though - perhaps we can reevaluate in the parent PR?

diego-plan9 commented on qiskit/wrapper/_wrapper.py:
```diff
@@ -38,17 +37,33 @@ def register(token, url='https://quantumexperience.ng.bluemix.net/api',
+        'token': token, 'url': url, 'hub': hub, 'group': group,
+        'project': project, 'proxies': proxies, 'verify': verify
+    }
+    _DEFAULT_PROVIDER.add_ibmq_provider(credentials, provider_name)
```
Related to the comment at https://github.com/QISKit/qiskit-core/pull/584#discussion_r197174310 - I went for being conservative when exposing the providers to the user in the wrapper, but happy to reevaluate eventually.

diego-plan9 commented:
Thanks for the review, @delapuente ! I have addressed most of the issues, keeping some of them still open with the idea of further discussing and addressing them when #540 is tackled while still providing a functional `unregister()`.

delapuente commented on qiskit/wrapper/defaultqiskitprovider.py:
```diff
@@ -103,6 +110,16 @@ def add_provider(self, provider, provider_name):
 
+        # Check for backend name clashes, emitting a warning.
+        current_backends = {str(backend) for backend in self.available_backends()}
+        added_backends = {str(backend) for backend in provider.available_backends()}
```
Non-blocking: can we make it clear with the name that we did not add the backends yet? That they are `incoming_backends` or `new_provider_backends` or something like this?

delapuente commented on qiskit/wrapper/defaultqiskitprovider.py:
```diff
@@ -87,6 +88,12 @@ def add_provider(self, provider, provider_name):
         Add a new provider to the list of known providers.
 
+        Note:
+            If a backend from the new provider has a name that is in use by an
```
Non-blocking but what about: `If some backend of the new provider has a name in use by an already registered provider, the backend will not be available, and the name will always refer to that previously registered.`

delapuente commented on qiskit/wrapper/defaultqiskitprovider.py:
```diff
@@ -103,6 +110,16 @@ def add_provider(self, provider, provider_name):
+        if common_backends:
+            logger.warning(
+                'The backend names "%s" (provided by "%s") are already in use. '
+                'Consider using unregister() for avoiding namespace conflicts.',
```
Non-blocking but I would seriously consider changing `namespace` with simply `name`.

delapuente commented on test/python/test_wrapper.py:
```diff
@@ -92,6 +93,28 @@ def test_unregister_non_existent(self):
+                                provider_name='provider1')
+        initial_providers = registered_providers()
+        initial_backends = qiskit.wrapper.available_backends()
+        with self.assertLogs(level=logging.WARNING) as context:
```
Can we rename `context` with `logger_output` or simply `output`? Does it make sense?

delapuente commented on test/python/test_wrapper.py:
```diff
@@ -92,6 +93,28 @@ def test_unregister_non_existent(self):
+        self.assertCountEqual(initial_providers + ['provider2'],
+                              registered_providers())
+        # Check that no new backends have been added.
+        self.assertCountEqual(initial_backends,
```
We should also assert that the backend name still refers to the already registered backend:

```python
initial_backends = qiskit.wrapper.available_backends()
original_backend = qiskit.wrapper.get_backend('ibmqx4')
# ...
with self.assertLogs(level=logging.WARNING) as context:
  qiskit.wrapper.register(QE_TOKEN, QE_URL, hub, group, project, provider_name='provider2')
# ...
self.assertEqual(original_backend, qiskit.wrapper.get_backend('ibmqx4'))
```
